### ISSUE TITLE
ignoring lint errors

### ISSUE THREAD

**Issue by BarnumD** [lingua-pupuli/puppet-vscode#176]
In [this documentation](http://puppet-lint.com/controlcomments/), You can add something like `# lint:ignore:140chars` to the end of a line to have that particular line ignored by the lint parser.  This seems to work fine with the puppet-lint gem built into our CI pipeline.  However, in vscode with this extension it still complains about the line.

**Comment by BarnumD** [comment]
I have VSCode 1.17.2, puppet-vscode extension 0.7.1.  Puppet version in powershell is 4.8.0.  The version listed in the bottom of vscode when the extension launches is 4.7.0.

I have a file with `# lint:ignore:140chars` and ` # lint:endignore` on separate lines.  That seems to be working properly with the code in between.

What seems to be acting strangely is when you have single line ignores.  I have single line ignores in other files.  For instance, the line below displays a 'lint has more than 140 characters warning'
`command  => 'docker images | grep -v REPOSITORY| grep -v "<none>" | awk \'{print $1":"$2}\'| xargs -n1 docker pull', # lint:ignore:140chars`

If you put two of the same lines back to back, you only get one warning (I would expect to get none.. or two (erroneously) .. but one?)  If you put something else in the middle of them like `user     => 'root',` or any other line then you get two warnings.

**Comment by glennsarti** [comment]
@BarnumD Thanks for the error report.  I'll see if can repro it

**Comment by glennsarti** [comment]
Hi @BarnumD I'm not able to reproduce the issue. Here's what I have:

No errors:
```
exec { 'Wee!':
  command => 'docker images | grep -v REPOSITORY| grep -v "<none>" | awk \'{print $1":"$2}\'| xargs -n1 docker pull',
}

exec { 'Wee!':
  command => 'docker images | grep -v REPOSITORY| grep -v "<none>" | awk \'{print $1":"$2}\'| xargs -n1 docker pull docker images | grep -v REPOSITORY| grep -v "<none>" | awk \'{print $1":"$2}\'| xargs -n1 docker pull ', # lint:ignore:140chars
}
```

Lint errors
Line has more than 140 characters (Line 6, Char 140)
```
exec { 'Wee!':
  command => 'docker images | grep -v REPOSITORY| grep -v "<none>" | awk \'{print $1":"$2}\'| xargs -n1 docker pull',
}

exec { 'Wee!':
  command => 'docker images | grep -v REPOSITORY| grep -v "<none>" | awk \'{print $1":"$2}\'| xargs -n1 docker pull docker images | grep -v REPOSITORY| grep -v "<none>" | awk \'{print $1":"$2}\'| xargs -n1 docker pull ',
}
```

Lint Errors
message: 'line has more than 140 characters'
at: '10,140'
```
exec { 'Wee!':
  command => 'docker images | grep -v REPOSITORY| grep -v "<none>" | awk \'{print $1":"$2}\'| xargs -n1 docker pull',
}

exec { 'Wee!':
  command => 'docker images | grep -v REPOSITORY| grep -v "<none>" | awk \'{print $1":"$2}\'| xargs -n1 docker pull docker images | grep -v REPOSITORY| grep -v "<none>" | awk \'{print $1":"$2}\'| xargs -n1 docker pull ', # lint:ignore:140chars
}

exec { 'Wee!':
  command => 'docker images | grep -v REPOSITORY| grep -v "<none>" | awk \'{print $1":"$2}\'| xargs -n1 docker pull docker images | grep -v REPOSITORY| grep -v "<none>" | awk \'{print $1":"$2}\'| xargs -n1 docker pull ',
}
```

**Comment by glennsarti** [comment]
Are you able to post a more complete example of the manifest and the lint errors you're seeing.

**Comment by BarnumD** [comment]
Take the three of these together..
```
exec { 'Wee!':
  command => 'echo The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog.', # lint:ignore:140chars
}

exec { 'Wee!':
  command => 'echo The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog.',
}

exec { 'Wee!':
  command  => 'echo The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog.', # lint:ignore:140chars
}
```

It should produce a warning for only the second instance.  In my vscode it produces an error on the second and third.  Interestingly, if you take away the first the warning on the third goes away.  It seems like it doesn't error on the first instance of `#lint`, but causes issues after that.

**Comment by glennsarti** [comment]
Hrmm ok...I can't repro it in puppet 5.x series.  Time to try 4.7.  Thanks @BarnumD

**Comment by glennsarti** [comment]
Nope not puppet version related.

![image](https://user-images.githubusercontent.com/5819622/32302023-cb105b76-bf1d-11e7-90cf-d4bbc01d0d50.png)

**Comment by BarnumD** [comment]
Just me then :(
![image](https://user-images.githubusercontent.com/2438253/32337137-8f812c52-bfc7-11e7-85f9-670126ecace7.png)

**Comment by glennsarti** [comment]
@jpogran ++ for the keen eyes.  Looks like puppet-lint doesn't like CRLF line endings.  Try using LF line endings.  I can repro your errors if I change that file to CRLF liek you have.

**Comment by glennsarti** [comment]
@BarnumD ^^

**Comment by BarnumD** [comment]
Yes, on linux nodes.

**Comment by BarnumD** [comment]
LF line endings did fix the extra warnings.  Should I leave this open or close it if it's not really an issue for puppet-vscode?

**Comment by glennsarti** [comment]
I'm happy to close it.  And I can raise an issue on puppet-lint and cross reference.

**Comment by glennsarti** [comment]
Hrmm...this may not be a puppet-lint issue yet.  puppet-lint still behaves correctly with CRLF

**Comment by glennsarti** [comment]
Well..this has turned into quite a rabbit hole...

So puppet-lint loads the file using text mode, not binary mode.
https://github.com/rodjek/puppet-lint/blob/0799152d9ded067e1a3c5b2fa4f2748a30fbf2c5/lib/puppet-lint.rb#L78-L80

This means all of the CRLFs are magically turned into LFs when using puppet-lint via the command line.

The Language Server however holds the file in memory as string and is a direct copy of what you type in the editor.

Changing puppet-lint at https://github.com/rodjek/puppet-lint/blob/0799152d9ded067e1a3c5b2fa4f2748a30fbf2c5/lib/puppet-lint.rb#L78-L80 to use binary mode `rb:UTF-8` makes puppet lint throw the same errors as VS Code.

Because Puppet manifests can be either LF or CRLF, puppet doesn't care therefore puppet-lint shouldn't care.  The question is, is this a puppet-lint error? 

Clearly puppet-lint can't handle CRLFs correctly, but it's never had to due to the file loading behaviour.

Now I can emulate this in Language Server by just converting CRLF to LF prior to linting, but this could become expensive as every keystroke (almost) would cause a full text search on linting.

**Comment by glennsarti** [comment]
Cross referenced - https://github.com/rodjek/puppet-lint/issues/778

**Comment by glennsarti** [comment]
Looks like puppet-lint may have a fix being worked on.  https://github.com/rodjek/puppet-lint/pull/782

**Comment by glennsarti** [comment]
Pending a puppet-lint release with commit https://github.com/rodjek/puppet-lint/commit/2a850ab3fd3694a4dd0c4d2f22a1e60b9ca0a495 in it.

**Comment by glennsarti** [comment]
Still waiting for a release of https://github.com/rodjek/puppet-lint/pull/782

**Comment by glennsarti** [comment]
The fix was released in 2.3.5 and verified it's now fixed in the latest extension release 0.10.0
